test/boot_raw_fallback.c

```diff
@@ -0,0 +1,414 @@
+		return;
+
+	/* prepare slot */
+	slotpath = g_strdup(g_getenv("RAUC_TEST_BLOCK_LOOP"));
+	if (!slotpath) {
+		g_test_message("no block device for testing found (define RAUC_TEST_BLOCK_LOOP)");
+		g_test_skip("RAUC_TEST_BLOCK_LOOP undefined");
+		return;
+	}
+
+	slot_size = get_file_size(slotpath, &ierror);
+	g_assert_no_error(ierror);
+	g_assert(slot_size != 0);
+
+	fill_file(slotpath, 0, slot_size, 0x00, FALSE);
+
+	header_size = sector_size;
+	if (data->image_size < header_size)
+		header_size = data->image_size;
+
+	/* pre-install parts of primary location */
+	if (data->options & OPT_PRE_INSTALL_PRIMARY_HEAD)
+		g_assert_true(fill_file(slotpath, data->region_start, header_size, 0x00, TRUE));
+
+	if (data->options & OPT_PRE_INSTALL_PRIMARY_IMG) {
+		g_assert_true(fill_file(slotpath, data->region_start, data->image_size, 0x00, TRUE));
+		if (!(data->options & OPT_PRE_INSTALL_PRIMARY_HEAD))
+			g_assert_true(fill_file(slotpath, data->region_start, header_size, 0x00, FALSE));
+	}
+
+	/* pre-install parts of fallback location */
+	if (data->options & OPT_PRE_INSTALL_FALLBACK_HEAD)
+		g_assert_true(fill_file(slotpath, data->region_start + (data->region_size / 2), header_size, 0x00, TRUE));
+
+	if (data->options & OPT_PRE_INSTALL_FALLBACK_IMG) {
+		g_assert_true(fill_file(slotpath, data->region_start + (data->region_size / 2), data->image_size, 0x00, TRUE));
+		if (!(data->options & OPT_PRE_INSTALL_FALLBACK_HEAD))
+			g_assert_true(fill_file(slotpath, data->region_start + (data->region_size / 2), data->image_size, 0x00, FALSE));
+	}
+
+	/* pre-fill primary partition with dummy data */
+	if (data->options & OPT_DUMMY_FILL_PRIMARY)
+		g_assert_true(fill_file(slotpath, data->region_start, (data->region_size / 2), 0xAA, FALSE));
+
+	/* pre-fill fallback location with dummy data */
+	if (data->options & OPT_DUMMY_FILL_FALLBACK)
+		g_assert_true(fill_file(slotpath, data->region_start + (data->region_size / 2), (data->region_size / 2), 0xAA, FALSE));
+
+	image = create_source_image(fixture->tmpdir, "image.img", data->image_size);
+	g_assert_nonnull(image);
+
+	/* create target slot */
+	targetslot = g_new0(RaucSlot, 1);
+	targetslot->name = g_intern_string("bootloader.0");
+	targetslot->sclass = g_intern_string("bootloader");
+	targetslot->device = g_strdup(slotpath);
+	targetslot->type = g_strdup("boot-raw-fallback");
+	targetslot->region_start = data->region_start;
+	targetslot->region_size = data->region_size;
+	targetslot->state = ST_INACTIVE;
+
+	/* Set mount path to current temp dir */
+	mountprefix = g_build_filename(fixture->tmpdir, "testmount", NULL);
+	g_assert_nonnull(mountprefix);
+	r_context_conf()->mountprefix = mountprefix;
+	r_context();
+	g_assert(g_mkdir(mountprefix, 0777) == 0);
+
+	/* get handler for this */
+	handler = get_update_handler(image, targetslot, &ierror);
+	g_assert_no_error(ierror);
+	g_assert_nonnull(handler);
+
+	/* Run to perform an update */
+	res = handler(image, targetslot, hookpath, &ierror);
+
```

Comment:
You could wrap this with some `g_test_expect_message()` before and `g_test_assert_expected_messages()` after the handler call. This way, you can check if the "Updating primary partition at 0x??? on *" and "Updating fallback partition at 0x??? on *" appear in the correct order.

Reply:
Nice! That is a very good idea.
Thanks for this tip!